File: dockerfiles/turtlebot3/src/py_example_package/py_example_package/dummy_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class DummyPublisher(Node):
    def __init__(self):
        super().__init__('dummy_publisher')

        # Declare parameters
        self.declare_parameter('topic_name', 'chatter')
        self.declare_parameter('publish_rate', 1.0)
        self.declare_parameter('message', 'Hello from dummy_publisher')

        # Get parameters
        topic_name = self.get_parameter('topic_name').get_parameter_value().string_value
        self.publish_rate = self.get_parameter('publish_rate').get_parameter_value().double_value
        self.message = self.get_parameter('message').get_parameter_value().string_value

        # Set up publisher and timer
        self.publisher_ = self.create_publisher(String, topic_name, 10)
        self.timer = self.create_timer(1.0 / self.publish_rate, self.publish_message)

    def publish_message(self):
        msg = String()
        msg.data = self.message
        self.get_logger().info(f"Publishing: {msg.data}")
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = DummyPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: dockerfiles/turtlebot3/src/c_example_package/src/dummy_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class DummyPublisher : public rclcpp::Node
{
public:
    DummyPublisher()
    : Node("dummy_publisher")
    {
        // Declare parameters
        this->declare_parameter<std::string>("topic_name", "chatter");
        this->declare_parameter<double>("publish_rate", 1.0);
        this->declare_parameter<std::string>("message", "Hello from dummy_publisher");

        // Get parameters
        topic_name_ = this->get_parameter("topic_name").as_string();
        publish_rate_ = this->get_parameter("publish_rate").as_double();
        message_ = this->get_parameter("message").as_string();

        // Create publisher
        publisher_ = this->create_publisher<std_msgs::msg::String>(topic_name_, 10);

        // Create timer
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(static_cast<int>(1000 / publish_rate_)),
            std::bind(&DummyPublisher::publish_message, this));
    }

private:
    void publish_message()
    {
        auto message = std_msgs::msg::String();
        message.data = message_;
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);
    }

    std::string topic_name_;
    double publish_rate_;
    std::string message_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DummyPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
